### ISSUE TITLE
`tokio::sync::Mutex::lock()` never awaited

### ISSUE THREAD

**Issue by niklasf** [rust-lang/rust-clippy#9823]
### Description

Recently I forgot to add `.await` when switching to `tokio::sync::Mutex`:

```rust
fn with_standard(mutex: &std::sync::Mutex<()>) {
    let guard = mutex.lock().unwrap();
    // Critical section
    drop(guard); // Or change binding above to _guard
}

async fn with_tokio(mutex: &tokio::sync::Mutex<()>) {
    let guard = mutex.lock(); // async fn lock() never awaited!
    // Critical section
    drop(guard); // Or change binding above to _guard
}
```

Unfortunately, when the mutex does not carry data, it's easy to miss that `guard` does not have the expected type (a future, rather than a lock guard).

Would detecting one of these make sense?

* This pattern for `tokio::sync::Mutex` specifically.
* Future unused except drop. Though edge cases may exist where a future is created just to drop it.
* Future returned from `async fn` unused except drop. This seems to be always useless.

### Additional Labels

@rustbot label +A-lint

**Comment by llogiq** [comment]
I think this doesn't need a lint anymore as the `Future` trait is marked `#[must_use]`, so this will already raise a warning.
